[Sketcher] Round 1 of refactors

AjinkyaDahale opened the pull request:
Done as [a grant by the FPA](https://github.com/FreeCAD/FPA-grant-proposals/issues/6).

Adds some tests for `split` and `trim` functions in `SketchObject`, and refactors them. Plenty of changes are still needed after this, especially in `trim`: not all necessary tests are added, and the cognitive complexity is only reduced from 233 to 92. However, the many corner cases and constraint combinations in `trim` make this a significantly difficult task. The method has been almost completely rewritten, so this needs thorough testing.

3x380V commented:
Does „[Sketcher] Refactor SketchObject::trim()” belong here or not (either commit message is wrong or it should be omited).
Also no `else if` after `return`. I guess you will touch such code again, but still... Nice job, btw.

AjinkyaDahale commented:
> Does „[Sketcher] Refactor SketchObject::trim()” belong here or not (either commit message is wrong or it should be omited).

It is similar to commits that refactor other methods. I don't see how it doesn't belong.

> Also no `else if` after `return`. I guess you will touch such code again, but still... Nice job, btw.

Could you point to the line of code? There are many cases where I reduce indentation by using "if (...) {return;} ..." instead of using else.

FEA-eng commented:
Is it planned for 1.0 or post 1.0? It could be risky if done for 1.0.

3x380V commented:
> > Does „[Sketcher] Refactor SketchObject::trim()” belong here or not (either commit message is wrong or it should be omited).
> 
> It is similar to commits that refactor other methods. I don't see how it doesn't belong.

It's commit message reads _DO NOT MERGE_:
```
Sketcher] Refactor SketchObject::trim()

DO NOT MERGE because not complete: not all constraints are not transferred.

Includes:

* Adjust `deriveConstraintsForPieces` for trim
* Make sure new points are constrained on the cutting edges
* Add constraints at trim end(s) dependent on pre-exisiting ones with cutting curve(s). Also change tests to reflect that auto-coincidence can now be applied to non-periodic curves as well.

Known issue(s):
* Some redundant constraints may be added under certain circumstances (e.g. in
presence of perpendicular-via-point)
```
Are these known issues introduced by this commit?

> > Also no `else if` after `return`. I guess you will touch such code again, but still... Nice job, btw.
> 
> Could you point to the line of code? There are many cases where I reduce indentation by using "if (...) {return;} ..." instead of using else.

This part:  https://github.com/FreeCAD/FreeCAD/pull/15811/commits/91e2252f65f3ac1bc1e42241694b681051dcf734#diff-b9dbc09c2c4b8417d6df9b1640fe0b24441cf026658dd5220d8917aa1ebcf356R1515

AjinkyaDahale commented:
> It's commit message reads _DO NOT MERGE_:

The message is wrong. I will edit it shortly.
 
> Are these known issues introduced by this commit?

Yes, though the existing trim code itself probably has unknown issues. Just before starting this refactor, I fixed some in #14461.

3x380V commented:
Thank you. It just was not clear from commit message whenever this is existing issue or new regression. Also, I second @FEA-eng's question. Is this intended for 1.0?

AjinkyaDahale commented:
> Is it planned for 1.0 or post 1.0?

As I mentioned in the above comment, older trim code itself may have issues. One heuristic we could use is to check if any are reported. If yes, it is better to merge these changes (with any fixes) than to maintain older code.

That said, I'm OK with marking it for post 1.0.

AjinkyaDahale commented:
The failing test is related to a topo-naming test in PD, and likely broke because of the way trim now operates (making new curves rather than shrinking the old one). The fix for the test just before that is already in this commit, and I'm trying to fix this too now.

AjinkyaDahale commented:
Test now passes after I changed the behaviour of of new objects in trim back to the old one. However
1. There are still some fixme's left. I don't understand how the new curves are both construction if the original one is too, when I didn't explicitly make them so. Basically code is working when I'd expect it to not.
2. This change may also be needed in Split.

As such this PR may be made a draft, but I need some eyes on it to tell me what it breaks. There's likely (still) plenty.

chennes commented on src/Mod/Sketcher/App/GeometryFacade.cpp:
```diff
@@ -67,17 +68,17 @@ std::unique_ptr<GeometryFacade> GeometryFacade::getFacade(Part::Geometry* geomet
-    // make_unique has no access to private constructor
-    // return std::make_unique<const GeometryFacade>(geometry);
-}
+// std::unique_ptr<const GeometryFacade> GeometryFacade::getFacade(const Part::Geometry* geometry)
+// {
+//     if (geometry) {
+//         return std::unique_ptr<const GeometryFacade>(new GeometryFacade(geometry));
+//     }
+//     else {
+//         return std::unique_ptr<const GeometryFacade>(nullptr);
+//     }
+//     // make_unique has no access to private constructor
+//     // return std::make_unique<const GeometryFacade>(geometry);
+// }
```
If we're leaving this code here there should be an explanation of why.

AjinkyaDahale replied:
Does literally the same thing as another override. Probably safe to delete (for now).

kadet1090 commented on src/Mod/Part/App/Geometry.cpp:
```diff
@@ -793,6 +793,11 @@ GeomBSplineCurve* GeomCurve::toNurbs(double first, double last) const
 
+GeomCurve* GeomCurve::createArc([[maybe_unused]] double first, [[maybe_unused]] double last) const
+{
+    return nullptr;
```
If this is intentional please leave a comment why it only returns nullptr. Maybe it should be abstract function instead? 

This is not a function that I'd expect to return nullptr in any scenario - if it is expected to fail it should use `std::optional` if not - it should be abstract (`=0`).

AjinkyaDahale replied:
I'm hesitant to use `=0` since there is not other abstract method in GeomCurve.

kadet1090 replied:
I'm not sure how this class is used, but returning `nullptr` here could cause bugs if someone does not know that such method must be implemented. If not `=0` maybe exception? Throwing exception seems better than causing segfault.

kadet1090 commented on src/Mod/Sketcher/App/GeometryFacade.h:
```diff
@@ -118,8 +118,9 @@ class SketcherExport GeometryFacade: public Base::BaseClass, private ISketchGeom
-    static std::unique_ptr<const GeometryFacade> getFacade(const Part::Geometry* geometry);
+    static std::unique_ptr<GeometryFacade> getFacade(const Part::Geometry* geometry,
+                                                     bool owner = false);
+    // static std::unique_ptr<const GeometryFacade> getFacade(const Part::Geometry* geometry);
```
Same as previous - if it is commented for a reason it should be stated somewhere.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1375,68 +1376,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
+    else if (PosId == PointPos::end)
+        return aoe->getEndPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::mid)
+        return aoe->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfHyperbola *aoh, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aoh->getStartPoint();
+    else if (PosId == PointPos::end)
+        return aoh->getEndPoint();
+    else if (PosId == PointPos::mid)
+        return aoh->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfParabola *aop, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aop->getStartPoint();
+    else if (PosId == PointPos::end)
+        return aop->getEndPoint();
+    else if (PosId == PointPos::mid)
+        return aop->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomBSplineCurve *bsp, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return bsp->getStartPoint();
+    else if (PosId == PointPos::end)
+        return bsp->getEndPoint();
+    return Base::Vector3d();
+}
```
All these methods seems like good candidates for switch statements.

AjinkyaDahale replied:
Done.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1375,68 +1376,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
-            return bsp->getEndPoint();
+        return getPointForGeometry<Part::GeomPoint>(static_cast<const Part::GeomPoint*>(geo), PosId);
+    }
+    else if (geo->is<Part::GeomLineSegment>()) {
```
This is a minor suggestion, mostly loudly thinking. 

Maybe we could use something like

```
if (auto lineSegment = geo->as<Part::GeomLineSegment>()) {
    return getPointForGeometry<Part::GeomLineSegment>(lineSegment, PosId);
}
```

is the `<Part::GeomLineSegment>` even necessary for `getPointForGeometry`? It should infer it from argument, but I'm not that familiar with c++.

AjinkyaDahale replied:
Didn't know about 'geo->as<...>()`. It will definitely make it compact. However isn't having functions that has two effects in this fashion not good practice? Not to mention having an `=` in an if statement and treating it as a boolean. Is there a guarantee that this will evaluate to `false` iff the new variable is `nullptr`?

The redundant template is indeed something I can avoid, I just wasn't confident about it. With tests added maybe I can give it a try.

hyarion replied:
I think kadet's suggestion is really good here.

Regarding having `=` in an if statement, that's just modern c++. https://en.cppreference.com/w/cpp/language/if#if_statements_with_initializer

It's not more stranger than having a for loop with initializer.

And regarding the question if `nullptr` will evaluate to `false` - yes, it will in this type of context https://stackoverflow.com/questions/57527189/is-nullptr-falsy

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1593,6 +1656,23 @@ int SketchObject::addGeometry(std::unique_ptr<Part::Geometry> newgeo, bool const
 
+bool SketchObject::isClosedCurve(const Part::Geometry* geo)
+{
+    return (geo->is<Part::GeomCircle>()
+            || geo->is<Part::GeomEllipse>()
+            || (geo->is<Part::GeomBSplineCurve>()
+                && static_cast<const Part::GeomBSplineCurve*>(geo)->isPeriodic()));
```
combining || and && in singular expression is hard to read. Contents of inner () should be extracted to its own variable:

```c++
const auto isPeriodicBSpline = geo->is<Part::GeomBSplineCurve>() && static_cast<const Part::GeomBSplineCurve*>(geo)->isPeriodic());
```

and then used in result.

I'm also not a fan of using static_casts here - we could use `as` helper inside class geometry.

AjinkyaDahale replied:
Again, `as` is new to me. Will take a look.

kadet1090 replied:
It does not exist right now, but given the code I think that it would be quit useful to have as this contains a lot of casts.

AjinkyaDahale replied:
Sounds like a good candidate for an issue.

Coming to think of it, this may be exactly what dynamic cast does (and is probably what is being used in `geo->is`)/

kadet1090 replied:
Yes, geo->is uses `isDerivedFrom` afair which is also base for `Base::freecad_dynamic_cast` - you can see how I implemented `getObject<T>` here: https://github.com/FreeCAD/FreeCAD/pull/17528/commits/3640318ce58d2bdd385329beb06d5ec0ec637417. This would be very similar.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -3320,159 +3388,51 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
-            this->Constraints.setValues(std::move(newConstraints));
-            delConstraints(delete_list, false);
-        }
+        // if (!(constr->Type == Sketcher::Tangent
+        //       || constr->Type == Sketcher::Perpendicular)) {
+        //     return;
+        // }
+
+        // if (constr->First == GeoId1 && constr->Second == GeoId) {
+        //     constr->Type = constr->Type;
+        //     if (secondPos == Sketcher::PointPos::none)
+        //         secondPos = constr->FirstPos;
+        //     delete_list.push_back(constrId);
+        // }
+        // else if (constr->First == GeoId && constr->Second == GeoId1) {
+        //     constr->Type = constr->Type;
+        //     if (secondPos == Sketcher::PointPos::none)
+        //         secondPos = constr->SecondPos;
+        //     delete_list.push_back(constrId);
+        // }
```
same comment as other commented out code

AjinkyaDahale replied:
Removed. I want to keep this method in the future to remove complexity from elsewhere, so just left a `TODO` for now.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -3556,449 +3497,470 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
-    //******************* Step B.1 => Trimming for GeomTrimmedCurves (line segment and arcs)
-    //****************************************//
-    if (isDerivedFromTrimmedCurve || isNonPeriodicBSpline) {
+    double firstParam = static_cast<const Part::GeomCurve*>(geo)->getFirstParameter();
+    double lastParam = static_cast<const Part::GeomCurve*>(geo)->getLastParameter();
```
This `static_cast` is repeated - maybe it should be extracted to variable?

AjinkyaDahale replied:
Done.

kadet1090 commented on tests/src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -161,6 +263,944 @@ TEST_F(SketchObjectTest, testGeoIdFromShapeTypeRootPoint)
+    Part::GeomLineSegment lineSeg;
+    setupLineSegment(lineSeg);
+    // TODO: create curves intersecting at the right spots
+    Base::Vector3d trimPoint(lineSeg.pointAtParameter(
+        lineSeg.getFirstParameter()
+        + (lineSeg.getLastParameter() - lineSeg.getFirstParameter()) * 0.2));
```
This and below could use a better formatting. Maybe some function like `getPointAtNormalizedParam(double t)`?

AjinkyaDahale replied:
Done.

kadet1090 reviewed (changes requested):
Overall looks good! If I commented on something that I should not comment yet - feel free to remind me of it and ignore the comment.

3x380V commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1391,68 +1391,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfCircle *aoc, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aoc->getStartPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::end)
+        return aoc->getEndPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::mid)
+        return aoc->getCenter();
+    return Base::Vector3d();
```
FYI, @AjinkyaDahale commented on this previously in  https://github.com/FreeCAD/FreeCAD/issues/14696#issuecomment-2171811207

AjinkyaDahale replied:
OK. It's probably less advantageous to use switch statement when there are just 2-3 options but it's still a marginal clarity.

AjinkyaDahale commented:
4/5 new commits should be safely squashable. @kadet1090 can you confirm this addresses your concerns? 

The thing with the formatting `for` loops with single lines is that there are like 400 odd such changes to be made, so I'll need some sort of automation...

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1381,68 +1382,168 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
+        return lineSeg->getEndPoint();
+    }
+    default:
+        break;
+    }
+    return Base::Vector3d();
```
return could be inside the default

AjinkyaDahale replied:
I'd like to leave it outside just as a matter of preference to have some return outside.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1391,68 +1391,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
     return lastSolverStatus;
 }
 
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomPoint *geomPoint, PointPos PosId)
+{
+    if (PosId == PointPos::start || PosId == PointPos::mid || PosId == PointPos::end)
+        return geomPoint->getPoint();
+    return Base::Vector3d();
+}
```
```
// clang-format on
...
// clang-format off
```

AjinkyaDahale replied:
@chennes any input? Coming to think of it, changes may not all be self contained, and thus new code may not be re-formattable by itself. Still, a region-based clang-format could help out.

kadet1090 commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1461,10 +1562,12 @@ int SketchObject::getAxisCount() const
-    for (std::vector<Part::Geometry*>::const_iterator geo = vals.begin(); geo != vals.end(); geo++)
-        if ((*geo) && GeometryFacade::getConstruction(*geo)
-            && (*geo)->is<Part::GeomLineSegment>())
+    for (const auto& geo : vals) {
+        if (geo && GeometryFacade::getConstruction(geo)
+            && geo->is<Part::GeomLineSegment>()) {
             count++;
+        }
+    }
```
This seems like `std::count_if`

AjinkyaDahale replied:
Good catch. I have probably done it in a "later" PR in #14696, but can be done here as well.

kadet1090 reviewed (changes requested):
Few more comments - can be addressed later though.

The `return nullptr` one is blocking IMHO, forgetting to overload this method would result in crash which is no go. Either method should be forced to be implemented via `=0`, return default that does not cause crash and emit warning or throw exception.

hyarion commented:
I had totally missed this PR and started to refactoring sketcher.cpp myself yesterday night out of boredom.
I'm just going to leave it here. It's 100% untested but maybe you'll find something that you can use: https://github.com/FreeCAD/FreeCAD/commit/7596b71731d1860ec999f4938640ab53ce93ff16?diff=split

kadet1090 reviewed (approved):
I'm happy with current state. On merge meeting we agreed to replace `return nullptr` with exception - I'm approving it in advance as it is ok to be merged assuming that this small change is made.

AjinkyaDahale commented:
I'll try to make the `createArc` method virtual for geomcurve, since that's ideal. But as you can see from the build failures there are challenges.

AjinkyaDahale commented:
@PaddleStroke can you take a look at the create arc algorithm for offset curvez? I'll make a exception throwing stub for now.

PaddleStroke commented:
@AjinkyaDahale could you please point me to where this is please?

AjinkyaDahale commented:
Decided to go ahead and throw exception instead of making `GeomCurve::createArc` purely virtual, at least for now.

AjinkyaDahale commented:
> @AjinkyaDahale could you please point me to where this is please?

@PaddleStroke I was referring to `Part::GeomOffsetCurve`. I had pinged you since you have implemented offset in sketcher, so would be able to add the method trivially. But coming to think of it that was done without `GeomOffsetCurve`, right? Of course you're welcome to still give it a try.

To implement `createArc` for offset curves, we will need to know what the parameters actually mean, as well as probably a way to get the original curve from which the offset was made. This is probably a separate project by itself.

PaddleStroke commented on src/Mod/Sketcher/App/SketchObject.cpp:
```diff
@@ -1389,68 +1390,168 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomPoint *geomPoint, PointPos PosId)
+{
+    if (PosId == PointPos::start || PosId == PointPos::mid || PosId == PointPos::end)
```
Why for all PointPos but none you are returning the point, but for none you are returning Base::Vector3d()? How Base::Vector3d() is better than returning the point ?

hyarion replied:
@PaddleStroke selecting a range when commenting makes it easier to read on the PR page as you then can see all relevant lines directly in the comment on the PR (not just the selected line and three above) See step (5) on [adding-comments-to-a-pull-request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

For making it easier for the next person, here's a [link to the comment in the code](https://github.com/FreeCAD/FreeCAD/pull/15811/files/05448b980a35152eafe3990740a105b7fa556367#r1868916354).

AjinkyaDahale replied:
IIRC it is just consistent with what was being done earlier. That said most likely there should be an exception thrown (maybe with open/close principle applied).

Can you think of a case where `PointPos::None` could be useful with this method?

PaddleStroke replied:
It would just make the function more forgiving.
It's just that the function has a if to test if it's start end or mid. end or mid have not much more justification to be there than none. So I think we can just drop the if. If one querry the point geo the pos he gives should not matter.
